examples/static-transformer/Main.hs

```diff
@@ -116,23 +187,361 @@ data Dropout where
        }
     -> Dropout
 
+dropout :: Dropout -> Tensor dtype shape -> IO (Tensor dtype shape)
+dropout Dropout {..} = Torch.Static.Native.dropout dropoutProb dropoutTrain
+
 data TransformerLMLayer (dtype :: D.DType) (embedDim :: Nat) (numHeads :: Nat) (ffnDim :: Nat) where
   TransformerLMLayer
-    :: { selfAttention :: MultiheadAttention dtype embedDim numHeads
-       , fc0 :: Linear dtype embedDim ffnDim
-       , fc1 :: Linear dtype ffnDim embedDim
-       , ln0 :: LayerNorm dtype '[embedDim]
-       , ln1 :: LayerNorm dtype '[embedDim]
+    :: { tAttn :: MultiheadAttention dtype embedDim numHeads
+       , tAttnDropout :: Dropout
+       , tLN0 :: LayerNorm dtype '[embedDim]
+       , tLN1 :: LayerNorm dtype '[embedDim]
+       , tMLP :: TransformerLMMLP dtype embedDim ffnDim
        }
     -> TransformerLMLayer dtype embedDim numHeads ffnDim
 
+transformerLMLayer
+  :: forall dtype numHeads ffnDim embedDim headDim seqLen batchSize
+   . ( 1 <= numHeads
+     , embedDim ~ (headDim * numHeads)
+     , Mod (embedDim * 3) 3 ~ 0
+     , Div (embedDim * 3) 3 ~ embedDim
+     , KnownDType dtype
+     , All KnownNat [embedDim, numHeads, seqLen, batchSize, headDim]
+     , EndsWith '[seqLen, batchSize, embedDim] '[embedDim]
+     )
+  => TransformerLMLayer dtype embedDim numHeads ffnDim
+  -> Tensor 'D.Bool '[seqLen, batchSize]
+  -> Tensor dtype '[seqLen, batchSize, embedDim]
+  -> IO (Tensor dtype '[seqLen, batchSize, embedDim])
+transformerLMLayer TransformerLMLayer {..} keyPaddingMask input = do
+  (attn, _) <- multiheadAttention tAttn keyPaddingMask input
+  x         <- Main.dropout tAttnDropout attn
+  let x' = Main.layerNorm tLN0 (x `add` input)
+  x''       <- transformerLMMLP tMLP x'
+  return $ Main.layerNorm tLN1 (x'' `add` x')
+
+transformerLMLayer'
+  :: forall dtype numHeads ffnDim embedDim headDim seqLen batchSize
+   . ( 1 <= numHeads
+     , embedDim ~ (headDim * numHeads)
+     , Mod (embedDim * 3) 3 ~ 0
+     , Div (embedDim * 3) 3 ~ embedDim
+     , KnownDType dtype
+     , All KnownNat '[embedDim, numHeads, seqLen, batchSize, headDim]
+     , EndsWith '[seqLen, batchSize, embedDim] '[embedDim]
+     )
+  => TransformerLMLayer dtype embedDim numHeads ffnDim
+  -> Kleisli
+       IO
+       ( Tensor 'D.Bool '[seqLen, batchSize]
+       , Tensor dtype '[seqLen, batchSize, embedDim]
+       )
+       (Tensor dtype '[seqLen, batchSize, embedDim])
+transformerLMLayer' TransformerLMLayer {..} =
+  (arr snd &&& attn)
+    >>> uncurry add
+    ^>> Main.layerNorm tLN0
+    ^>> (id &&& transformerLMMLP' tMLP)
+    >>^ uncurry add
+    >>^ Main.layerNorm tLN1
```

Comment:
While this is a cute notation, I think that if we really want this library to become successfull then we should be very cautious about using a hundred different operators and `Kleisli`, because it's certainly not super beginner friendly.

Reply:
I agree! this was meant only as an illustration